### ISSUE TITLE
Resume after Behind-Live-Window-Exception issues

### ISSUE THREAD

**Issue by rajinikanth0602** [google/ExoPlayer#8675]
**Crashes on versions 2.13.+**

**Steps to Reproduce** :

1) Use HLS live stream URI as a data source. (Ex : https://cph-msl.akamaized.net/hls/live/2000341/test/master.m3u8)
2) Attach a PlaybackStatsListener to the exoplayer.
3) Reproducing `BehindLiveWindowException` will crash the app.  (Pause the live stream for more than twice the live window duration and then resume the video to reproduce `BehindLiveWindowException`)

Works fine if I don't attach the PlaybackStatsListener by removing the below line
`player?.addAnalyticsListener(playbackStatsListener)`

**StackTrace :** 

```
FATAL EXCEPTION: main
    java.lang.IllegalArgumentException
        at com.google.android.exoplayer2.util.Assertions.checkArgument(SourceFile:41)
        at com.google.android.exoplayer2.analytics.PlaybackStatsListener$PlaybackStatsTracker.updatePlaybackState(SourceFile:679)
        at com.google.android.exoplayer2.analytics.PlaybackStatsListener$PlaybackStatsTracker.onFinished(SourceFile:477)
        at com.google.android.exoplayer2.analytics.PlaybackStatsListener.onSessionFinished(SourceFile:183)
        at com.google.android.exoplayer2.analytics.DefaultPlaybackSessionManager.updateSessionsWithTimelineChange(SourceFile:182)
        at com.google.android.exoplayer2.analytics.PlaybackStatsListener.maybeAddSessions(SourceFile:304)
        at com.google.android.exoplayer2.analytics.PlaybackStatsListener.onEvents(SourceFile:252)
        at com.google.android.exoplayer2.analytics.AnalyticsCollector.lambda$setPlayer$1$AnalyticsCollector(SourceFile:141)
        at com.google.android.exoplayer2.analytics.-$$Lambda$AnalyticsCollector$zCA9RWls746W6T-nwroi1F_zK-4.invoke(Unknown Source:8)
        at com.google.android.exoplayer2.util.ListenerSet$ListenerHolder.iterationFinished(SourceFile:309)
        at com.google.android.exoplayer2.util.ListenerSet.handleMessage(SourceFile:253)
        at com.google.android.exoplayer2.util.ListenerSet.lambda$gt28PMFfhLXGfIVY4v7eP9kmalg(Unknown Source:0)
        at com.google.android.exoplayer2.util.-$$Lambda$ListenerSet$gt28PMFfhLXGfIVY4v7eP9kmalg.handleMessage(Unknown Source:2)
        at android.os.Handler.dispatchMessage(Handler.java:103)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7695)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)

```

**Code :**
```
fun initPlayer() {
    val bandwidthMeter: BandwidthMeter = DefaultBandwidthMeter
        .Builder(context)
        .build()

    val videoTrackSelectionFactory = AdaptiveTrackSelection.Factory()

    val trackSelector = DefaultTrackSelector(context, videoTrackSelectionFactory)
   
    player = SimpleExoPlayer.Builder(context)
        .setTrackSelector(trackSelector)
        .setBandwidthMeter(bandwidthMeter)
        .build()

    player?.addListener(playerEventListener)
    player?.addAnalyticsListener(playbackStatsListener)

    player?.prepare()
    player?.repeatMode = Player.REPEAT_MODE_ONE
    player?.playWhenReady = true

}

fun setMediaSource(url: String) {

    val dataSourceFactory: DataSource.Factory = DefaultHttpDataSourceFactory()

    val hlsMediaSource: HlsMediaSource = HlsMediaSource.Factory(dataSourceFactory)
        .setAllowChunklessPreparation(true)
        .createMediaSource(MediaItem.fromUri(url))

    player?.setMediaSource(hlsMediaSource, false)

    player?.seekToDefaultPosition()

    player?.playWhenReady = true

}

```

**Comment by tonihei** [comment]
I can't reproduce with just playing this stream, adding a `PlaybackStatsListener` and causing a `BehindLiveWindowException`. 

One thing I noted is that the code in the stack trace isn't triggered when an exception is thrown, but only later when something else happens (e.g. `player.release()`, `player.setMediaItems`, `player.stop(true)` or similar). Could you let us know what your app is doing once the `BehindLiveWindowExceptions` is reported?

**Comment by rajinikanth0602** [comment]
@tonihei we re-initialize the player when we get `BehindLiveWindowException`, `player?.prepare()` and `player?.setMediaSource` gets called in this process, it is working properly on <2.13.0.

**Code**

```
if (isBehindLiveWindow(error)) {
    try {
        reInitPlayer()
    } catch (e : Exception) {
        e.printStackTrace()
    }
}

fun reInitPlayer() {
    initPlayer()
    setMediaSource(url)
}
```


To reproduce the issue easily, change the minBuffer and maxBuffer to 5000, pause the video for long time and resume the video.

**Comment by tonihei** [comment]
Thanks for the reproduction steps. I can see the error as well now.

The reason for this is that you are registering the same listener to multiple players in parallel. When you call `reInitPlayer()`, you create a second `Player` (while the first one is still active) and register the same global listener to the second player as well. I think you should not create a second player and instead just call `player.seekToDefaultPosition() and player.prepare()` to recover from the `BehindLiveWindowException` as explained here: https://exoplayer.dev/live-streaming.html#behindlivewindowexception

This is still an issue though if you release and create a new player in one step, but keep using the same `PlaybackStatsListener`. I'll look into that.

**Comment by rajinikanth0602** [comment]
@tonihei https://exoplayer.dev/live-streaming.html#behindlivewindowexception this solution is not working for me, please check the player events below, it just seeks to the `live position` and tries to load the buffer forever.

**Version** : 2.13.2 

**Url** : https://cph-msl.akamaized.net/hls/live/2000341/test/master.m3u8

**Code :** 

```
if (isBehindLiveWindow(error)) {

    try {
        player?.seekToDefaultPosition()
        player?.prepare()
    } catch (e : Exception) {
        MLogger.e(TAG, "Error in reInitPlayer : ${e.message}" )
        e.printStackTrace()
    }
}
```

**Logs** 

```
2021-03-04 22:57:34.010 E: Playback error
com.google.android.exoplayer2.ExoPlaybackException: Source error
at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(SourceFile:579)
at android.os.Handler.dispatchMessage(Handler.java:103)
at android.os.Looper.loop(Looper.java:214)
at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: com.google.android.exoplayer2.source.BehindLiveWindowException
at com.google.android.exoplayer2.source.hls.HlsChunkSource.getNextChunk(SourceFile:396)
at com.google.android.exoplayer2.source.hls.HlsSampleStreamWrapper.continueLoading(SourceFile:708)
at com.google.android.exoplayer2.source.CompositeSequenceableLoader.continueLoading(SourceFile:80)
at com.google.android.exoplayer2.source.hls.HlsMediaPeriod.continueLoading(SourceFile:365)
at com.google.android.exoplayer2.source.MaskingMediaPeriod.continueLoading(SourceFile:226)
at com.google.android.exoplayer2.MediaPeriodHolder.continueLoading(SourceFile:217)
at com.google.android.exoplayer2.ExoPlayerImplInternal.maybeContinueLoading(SourceFile:2180)
at com.google.android.exoplayer2.ExoPlayerImplInternal.handleContinueLoadingRequested(SourceFile:2143)
at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(SourceFile:499)
at android.os.Handler.dispatchMessage(Handler.java:103) 
at android.os.Looper.loop(Looper.java:214) 
at android.os.HandlerThread.run(HandlerThread.java:67) 
2021-03-04 22:57:34.055 E: Event [onPlayerError : Source error]
2021-03-04 22:57:34.055 W:     at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(SourceFile:579)
2021-03-04 22:57:34.056 W:     at com.google.android.exoplayer2.ExoPlayerImplInternal.maybeContinueLoading(SourceFile:2180)
2021-03-04 22:57:34.056 W:     at com.google.android.exoplayer2.ExoPlayerImplInternal.handleContinueLoadingRequested(SourceFile:2143)
2021-03-04 22:57:34.056 W:     at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(SourceFile:499)
2021-03-04 22:57:34.056 D: Event -> isBehindLiveWindow
2021-03-04 22:57:34.056 E: isBehindLiveWindowError
2021-03-04 22:57:34.057 W: Event onIsLoadingChanged : false
2021-03-04 22:57:34.057 D: onIsLoadingChanged : false
2021-03-04 22:57:34.057 D: onIsLoadingChangedListener
2021-03-04 22:57:34.058 D: onIsLoadingChanged : false
2021-03-04 22:57:34.059 D: avgDownloadSpeedListener
2021-03-04 22:57:34.061 W: Event onPlaybackStateChanged : State 1
2021-03-04 22:57:34.061 D: onPlaybackStateChanged : 1
2021-03-04 22:57:34.061 D: onPlaybackStateChangedListener
2021-03-04 22:57:34.061 W: Event -> IDLE
2021-03-04 22:57:34.068 W: Event onIsPlayingChanged : false
2021-03-04 22:57:34.068 W: Event onPositionDiscontinuity : 1
2021-03-04 22:57:34.068 D: HasEndTag : false, Time Diff : 18000]
2021-03-04 22:57:34.070 W: Event onPlaybackStateChanged : State 2
2021-03-04 22:57:34.070 D: onPlaybackStateChanged : 2
2021-03-04 22:57:34.070 D: onPlaybackStateChangedListener
2021-03-04 22:57:34.070 W: Event -> BUFFERING
2021-03-04 22:57:34.074 W: Event onPositionDiscontinuity : 2
2021-03-04 22:57:34.074 D: HasEndTag : false, Time Diff : 18000]
2021-03-04 22:57:34.075 W: Event onIsLoadingChanged : true
2021-03-04 22:57:34.075 D: onIsLoadingChanged : true
2021-03-04 22:57:34.075 D: onIsLoadingChangedListener
2021-03-04 22:57:34.075 D: onIsLoadingChanged : true
2021-03-04 22:57:34.791 W: Event onTimelineChanged -> Offset : -48000
2021-03-04 22:57:34.792 D: HasEndTag : false, Time Diff : -48000]
2021-03-04 22:57:41.679 W: Event onTimelineChanged -> Offset : -42000
2021-03-04 22:57:41.679 D: HasEndTag : false, Time Diff : -42000]
2021-03-04 22:57:48.036 W: Event onTimelineChanged -> Offset : -36000
2021-03-04 22:57:48.036 D: HasEndTag : false, Time Diff : -36000]
2021-03-04 22:57:54.371 W: Event onTimelineChanged -> Offset : -30000
2021-03-04 22:57:54.371 D: HasEndTag : false, Time Diff : -30000]
2021-03-04 22:58:00.701 W: Event onTimelineChanged -> Offset : -24000
2021-03-04 22:58:00.701 D: HasEndTag : false, Time Diff : -24000]
2021-03-04 22:58:07.384 W: Event onTimelineChanged -> Offset : -18000
2021-03-04 22:58:07.385 D: HasEndTag : false, Time Diff : -18000]
```


**Working if we release the old player before re-initializing the new player.**

**Comment by tonihei** [comment]
Thanks for testing. This is indeed a separate issue independent of the one in `PlaybackStatsListener`.  The player attempts to seek to twice the default position far into the future and you need to wait until the actual time caught up before playback continues.

**Comment by hantrungkien** [comment]
@tonihei I'm also facing this issue. After Exo throws the BLWE then I call seekToDefaultPosition() and prepare() again but like the below image, currentDuration needs decrease to 0 to be able to continue resume live. So I have workarounded by setMediaSource() and call prepare() again but with this way the screen is blinking black. I expect resumable Exo with smooth way, is it possible?

![photo_2021-03-13_17-25-03](https://user-images.githubusercontent.com/20286370/111338705-b8071900-86a9-11eb-8aa1-cd9e78b16609.jpg)

**Comment by ojw28** [comment]
@tonihei - Is this bug fixed by the commit referenced above?

**Comment by tonihei** [comment]
No, the commit that fixed this isn't uploaded yet (internal ref cl/365558959)
